File: Data Structures & Algorithms/single-element-in-a-sorted-array/submission-1.py
class Solution:
    def singleNonDuplicate(self, nums: List[int]) -> int:
        # if len(nums) == 1: return nums[0]

        l, r = 0, len(nums) - 1
        # while l <= r:
        #     m = (l + r) // 2
        #     if l == r == m:
        #         return nums[m]
        #     if m%2 == 0:
        #         if nums[m] == nums[m + 1]:
        #             l = m
        #         else:
        #             r = m
        #     else:
        #         if nums[m] == nums[m + 1]:
        #             r = m - 1
        #         else:
        #             l = m + 1
        while l < r:
            m = (l + r) // 2
            if m % 2 != 0:
                m = m - 1
            if nums[m] == nums[m + 1]:
                l = m + 2
            else:
                r = m
        return nums[l]

File: Data Structures & Algorithms/subarray-product-less-than-k/submission-1.rs
impl Solution {
    pub fn num_subarray_product_less_than_k(nums: Vec<i32>, k: i32) -> i32 {
        let mut res = 0;
        let mut l = 0;
        let mut product = 1;
        for r in 0..nums.len() {
            product *= nums[r];
            while l <= r && product >= k {
                product = product / nums[l];
                l += 1
            }
            res += r - l + 1;
        }
        res as i32
    }
}
